app/client/rest/tracking.ts

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {DeviceEventEmitter, Platform} from 'react-native';
+
+import {Events} from '@constants';
+import {t} from '@i18n';
+import {setServerCredentials} from '@init/credentials';
+import {getFormattedFileSize} from '@utils/file';
+import {logDebug, logInfo} from '@utils/log';
+import {semverFromServerVersion} from '@utils/server';
+
+import * as ClientConstants from './constants';
+import ClientError from './error';
+
+import type {APIClientInterface, ClientHeaders, ClientResponseMetrics, RequestOptions} from '@mattermost/react-native-network-client';
+
+type UrlData = {
+    count: number;
+    metrics?: ClientResponseMetrics;
+}
+
+type GroupData = {
+    activeCount: number;
+    startTime: number;
+    totalSize: number;
+    totalCompressedSize: number;
+    urls: Record<string, UrlData>;
+    completionFlag: boolean;
+    completionTimer?: NodeJS.Timeout;
+}
+
+export default class ClientTraking {
+    apiClient: APIClientInterface;
+    csrfToken = '';
+    requestHeaders: {[x: string]: string} = {};
+    serverVersion = '';
+    urlVersion = '/api/v4';
+    enableLogging = false;
+
+    requestGroups: Map<string, GroupData> = new Map();
+
+    constructor(apiClient: APIClientInterface) {
+        this.apiClient = apiClient;
+    }
+
+    setBearerToken(bearerToken: string) {
+        this.requestHeaders[ClientConstants.HEADER_AUTH] = `${ClientConstants.HEADER_BEARER} ${bearerToken}`;
+        setServerCredentials(this.apiClient.baseUrl, bearerToken);
+    }
+
+    setCSRFToken(csrfToken: string) {
+        this.csrfToken = csrfToken;
+    }
+
+    getRequestHeaders(requestMethod: string) {
+        const headers = {...this.requestHeaders};
+
+        if (this.csrfToken && requestMethod.toLowerCase() !== 'get') {
+            headers[ClientConstants.HEADER_X_CSRF_TOKEN] = this.csrfToken;
+        }
+
+        return headers;
+    }
+
+    initTrackGroup(groupLabel: string) {
+        if (!this.requestGroups.has(groupLabel)) {
+            this.requestGroups.set(groupLabel, {
+                activeCount: 0,
+                startTime: Date.now(),
+                totalSize: 0,
+                totalCompressedSize: 0,
+                urls: {},
+                completionFlag: false,
+            });
+        }
+    }
+
+    trackRequest(groupLabel: string, url: string, metrics?: ClientResponseMetrics) {
+        this.initTrackGroup(groupLabel);
+
+        const group = this.requestGroups.get(groupLabel)!;
+
+        if (group.urls[url]) {
+            group.urls[url].count += 1;
+        } else {
+            group.urls[url] = {
+                count: 1,
+                metrics,
+            };
+        }
```

Comment:
What do you think about using just the path instead of the entire url?

Maybe it will give us more duplicates if any?

```
 LOG  {"path": "/api/v4/users/me/preferences", "query": undefined}
 LOG  {"path": "/api/v4/users/me/teams", "query": undefined}
 LOG  {"path": "/api/v4/users/me/teams/members", "query": undefined}
 LOG  {"path": "/api/v4/users/me", "query": undefined}
 LOG  {"path": "/api/v4/users/me/status", "query": undefined}
 LOG  {"path": "/api/v4/users/me/teams/rcgiyftm7jyrxnma1osd8zswby/channels/members", "query": undefined}
 LOG  {"path": "/api/v4/users/me/teams/rcgiyftm7jyrxnma1osd8zswby/channels", "query": "include_deleted=true&last_delete_at=1733860858455"}
 LOG  {"path": "/api/v4/users/me/teams/rcgiyftm7jyrxnma1osd8zswby/channels/categories", "query": undefined}
```
```suggestion
        const [path] = url.split('?');

        if (group.urls[path]) {
            group.urls[path].count += 1;
        } else {
            group.urls[path] = {
                count: 1,
                metrics,
            };
        }
```

Reply:
They would not be really duplicates though.. I think at this time we can keep the url and if we want to try and refine more then we can do soemething like this